File: 6_tree/树状数组/经典题/维护个数与值域和的值域树状数组/F - Max Matrix-维护个数与前缀和.py
# https://atcoder.jp/contests/jsc2021/tasks/jsc2021_f

# 输入 n m q (≤2e5)，初始你有长为 n 的数组 a，长为 m 的数组 b，元素值都为 0，下标从 1 开始。
# 然后输入 q 个询问，每个询问形如 t x y (1≤y≤1e8)。
# t=1，表示把 a[x]=y；t=2，表示把 b[x]=y。
# 每次修改后，输出 ∑∑max(a[i],b[j])，这里 i 取遍 [1,n]，j 取遍 [1,m]。

# https://atcoder.jp/contests/jsc2021/submissions/36189511
# https://atcoder.jp/contests/jsc2021/submissions/36189772

# 提示 1：元素的顺序并没有那么重要。

# !提示 2：每次修改时，只需要知道另一个数组中，有多少个数比自己小（这些数的个数乘上自己），以及≥自己的数的和是多少。

# 最好写的应该是离散化+树状数组。
# 注意把 0 也加到离散化中。

# TODO
# Run time Error
from typing import List, Tuple


def maxMatrix(n: int, m: int, queries: List[Tuple[int, int, int]]) -> List[int]:
    nums1, nums2 = [0] * n, [0] * m
    sum1, sum2 = BIT1(int(1e8 + 10)), BIT1(int(1e8 + 10))
    count1, count2 = BIT1(int(1e8 + 10)), BIT1(int(1e8 + 10))
    count1.add(0, n)
    count2.add(0, m)
    cur = 0
    res = []
    for op, qi, qv in queries:
        curNums, curCount, curSum = (nums1, count1, sum1) if op == 2 else (nums2, count2, sum2)
        otherCount, otherSum = (count2, sum2) if op == 2 else (count1, sum1)

        # 移除旧值
        preNum = curNums[qi]
        # (as min/max)
        cur -= otherCount.query(preNum) * preNum + otherSum.queryRange(preNum + 1, int(1e8 + 10))
        cur += otherCount.query(qv) * qv + otherSum.queryRange(qv + 1, int(1e8 + 10))
        curCount.add(preNum, -1)
        curSum.add(preNum, -preNum)
        # 加入新值
        curNums[qi] = qv
        curCount.add(qv, 1)
        curSum.add(qv, qv)

        res.append(cur)

    return res


if __name__ == "__main__":

    import sys

    class BIT1:
        """单点修改"""

        __slots__ = "size", "bit", "tree"

        def __init__(self, n: int):
            self.size = n
            self.tree = dict()

        def add(self, index: int, delta: int) -> None:
            # assert index >= 1, 'index must be greater than 0'
            index += 1
            while index <= self.size:
                self.tree[index] = self.tree.get(index, 0) + delta
                index += index & -index

        def query(self, index: int) -> int:
            index += 1
            if index > self.size:
                index = self.size
            res = 0
            while index > 0:
                res += self.tree.get(index, 0)
                index -= index & -index
            return res

        def queryRange(self, left: int, right: int) -> int:
            return self.query(right) - self.query(left - 1)

    sys.setrecursionlimit(int(1e9))
    input = lambda: sys.stdin.readline().rstrip("\r\n")
    n, m, q = map(int, input().split())
    Q = []
    for _ in range(q):
        op, qi, qv = map(int, input().split())
        Q.append((op, qi - 1, qv))
    print(*maxMatrix(n, m, Q), sep="\n")

File: 6_tree/树状数组/经典题/维护个数与值域和的值域树状数组/F - Max Matrix-维护个数与前缀和.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

// # TODO
// # Run time Error
func maxMatrix(n, m int, queries [][3]int) []int {
	nums1, nums2 := make([]int, n), make([]int, m)
	sum1, sum2 := CreateBIT(int(1e8+10)), CreateBIT(int(1e8+10))
	count1, count2 := CreateBIT(int(1e8+10)), CreateBIT(int(1e8+10))

	count1.Add(0, 0, n)
	count2.Add(0, 0, m)
	cur := 0
	res := make([]int, 0, len(queries))
	for _, q := range queries {
		op, qi, qv := q[0], q[1], q[2]
		var curNums []int
		var curCount, curSum BIT
		var otherCount, otherSum BIT
		if op == 2 {
			curNums, curCount, curSum = nums1, count1, sum1
			otherCount, otherSum = count2, sum2
		} else {
			curNums, curCount, curSum = nums2, count2, sum2
			otherCount, otherSum = count1, sum1
		}

		// 移除旧值
		preNum := curNums[qi]
		// (as min/max)
		cur -= otherCount.Query(0, preNum)*preNum + otherSum.Query(preNum+1, int(1e8+10))
		cur += otherCount.Query(0, qv)*qv + otherSum.Query(qv+1, int(1e8+10))
		curCount.Add(preNum, preNum, -1)
		curSum.Add(preNum, preNum, -preNum)
		// 加入新值
		curNums[qi] = qv
		curCount.Add(qv, qv, 1)
		curSum.Add(qv, qv, qv)

		res = append(res, cur)
	}

	return res

}

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, m, q int
	fmt.Fscan(in, &n, &m, &q)
	Q := make([][3]int, q)
	for i := range Q {
		fmt.Fscan(in, &Q[i][0], &Q[i][1], &Q[i][2])
		Q[i][1]--
	}

	res := maxMatrix(n, m, Q)
	for _, v := range res {
		fmt.Fprintln(out, v)
	}
}

func CreateBIT(n int) BIT {
	if n <= int(1e6) {
		return newSliceBIT(n)
	}

	return newMapBIT(n)
}

type BIT interface {
	// 区间 [left, right] 里的每个数增加 delta
	//  1 <= left <= right <= n
	Add(left, right, delta int)

	// 查询区间 [left, right] 的和
	Query(left, right int) int
}

type mapBIT struct {
	n     int
	tree1 map[int]int
	tree2 map[int]int
}

type sliceBIT struct {
	n     int
	tree1 []int
	tree2 []int
}

func newMapBIT(n int) *mapBIT {
	return &mapBIT{
		n:     n,
		tree1: make(map[int]int, 1<<10),
		tree2: make(map[int]int, 1<<10),
	}
}

func newSliceBIT(n int) *sliceBIT {
	return &sliceBIT{
		n:     n,
		tree1: make([]int, n+1),
		tree2: make([]int, n+1),
	}
}

func (bit *mapBIT) Add(left, right, delta int) {
	left, right = left+1, right+1
	bit.add(left, delta)
	bit.add(right+1, -delta)
}

func (bit *sliceBIT) Add(left, right, delta int) {
	left, right = left+1, right+1
	bit.add(left, delta)
	bit.add(right+1, -delta)
}

func (bit *mapBIT) Query(left, right int) int {
	left, right = left+1, right+1
	return bit.query(right) - bit.query(left-1)
}

func (bit *sliceBIT) Query(left, right int) int {
	left, right = left+1, right+1
	return bit.query(right) - bit.query(left-1)
}

func (bit *mapBIT) add(index, delta int) {
	if index <= 0 {
		errorInfo := fmt.Sprintf("index must be greater than 0, but got %d", index)
		panic(errorInfo)
	}

	rawIndex := index
	for index <= bit.n {
		bit.tree1[index] += delta
		bit.tree2[index] += (rawIndex - 1) * delta
		index += index & -index
	}
}

func (bit *sliceBIT) add(index, delta int) {
	if index <= 0 {
		errorInfo := fmt.Sprintf("index must be greater than 0, but got %d", index)
		panic(errorInfo)
	}

	rawIndex := index
	for index <= bit.n {
		bit.tree1[index] += delta
		bit.tree2[index] += (rawIndex - 1) * delta
		index += index & -index
	}
}

func (bit *mapBIT) query(index int) int {
	if index > bit.n {
		index = bit.n
	}

	rawIndex := index
	res := 0
	for index > 0 {
		res += rawIndex*bit.tree1[index] - bit.tree2[index]
		index -= index & -index
	}
	return res
}

func (bit *sliceBIT) query(index int) int {
	if index > bit.n {
		index = bit.n
	}

	rawIndex := index
	res := 0
	for index > 0 {
		res += rawIndex*bit.tree1[index] - bit.tree2[index]
		index -= index & -index
	}
	return res
}

func (bit *mapBIT) String() string {
	return "not implemented"
}

func (bit *sliceBIT) String() string {
	nums := make([]int, bit.n+1)
	for i := 0; i < bit.n; i++ {
		nums[i+1] = bit.Query(i+1, i+1)
	}
	return fmt.Sprint(nums)
}
